src/Tuple.js

```diff
@@ -0,0 +1,73 @@
+function Tuple(x, y) {
+    switch (arguments.length) {
+        case 0:
+            throw new TypeError('no arguments to Tuple');
+        case 1:
+            return function(y) {
+                return new _Tuple(x, y)
+            };
+        default:
+            return new _Tuple(x,y)
+    }
+};
+
+function _Tuple(x, y) {
+    this[0] = x;
+    this[1] = y;
+    this.length = 2;
+};
+
+function ensureConcat(xs) {
+    xs.forEach(function(x){
+      if(typeof x.concat != "function") {
+          var display = x.show ? x.show() : x;
+          throw new TypeError(display + " must be a semigroup to perform this operation");
+      };
+    })
+}
+
+Tuple.of = function(x) {
+    return Tuple(x, x);
+};
+
+Tuple.fst = function(x) {
+    return x[0];
+};
+
+Tuple.snd = function(x) {
+    return x[1];
+};
+
+_Tuple.prototype.of = function(x) {
+    return Tuple(this[0], x)
+}
+
+// semigroup
+_Tuple.prototype.concat = function(x) {
+    ensureConcat([this[0], this[1], x[0], x[1]])
```

Comment:
We're not testing that `this[0]` and `x[0]` have the same type, and would probably never get to that point.  So is there any good reason to even pass `x[0]` (or of course `x[1]`?)  It seems that we only need to check that the components of `this` can call `concat`.